File: app/src/main/java/com/mamba/picme/features/idphoto/IDPhotoViewModel.kt
@file:Suppress("TooGenericExceptionCaught") // 通用兜底：catch(Exception) 防崩溃，已记录日志
package com.mamba.picme.features.idphoto

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.provider.MediaStore
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mamba.picme.R
import com.mamba.picme.core.common.Logger
import com.mamba.picme.domain.matting.BackgroundComposer
import com.mamba.picme.domain.matting.BrushStroke
import com.mamba.picme.domain.matting.EdgeParams
import com.mamba.picme.domain.matting.IDPhotoComposer
import com.mamba.picme.domain.matting.IDPhotoSpecs
import com.mamba.picme.domain.matting.MaskPostProcessor
import com.mamba.picme.domain.matting.MaskSource
import com.mamba.picme.domain.matting.MattingEngine
import com.mamba.picme.domain.matting.MattingEngineImpl
import com.mamba.picme.domain.matting.StrokeLayer
import com.mamba.picme.domain.matting.StrokeMode
import com.mamba.picme.domain.matting.StrokePoint
import com.mamba.picme.domain.repository.MediaRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val TAG = "PoLang:IDPhoto"
private const val DECODE_MAX_DIM = 1024
private const val JPEG_QUALITY = 95
private const val MIN_ZOOM = 1f
private const val MAX_ZOOM = 4f

@Suppress("TooManyFunctions") // 页面 VM：底部 4-tab 的动作方法集中于此，拆分反而割裂状态一致性
class IDPhotoViewModel(
    private val mattingEngine: MattingEngine,
    private val mediaRepository: MediaRepository
) : ViewModel() {

    sealed class State {
        object Loading : State()
        data class Ready(
            val originalBitmap: Bitmap,
            val alpha: FloatArray,
            val alphaWidth: Int,
            val alphaHeight: Int,
            val subject: IDPhotoComposer.SubjectBounds? = null,
            val offsetX: Float = 0f,
            val offsetY: Float = 0f,
            val zoom: Float = 1f,
            val selectedColorIndex: Int = 0,
            val selectedSizeIndex: Int = 0,
            val activeTab: IdPhotoTab = IdPhotoTab.BG_COLOR,
            val edgeParams: EdgeParams = EdgeParams(),
            val strokeVersion: Int = 0,
            val canUndoStroke: Boolean = false,
            val canRedoStroke: Boolean = false,
            val isSaving: Boolean = false,
            val error: String? = null
        ) : State()
        data class Error(val message: String) : State()
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    private var appContext: Context? = null

    fun load(context: Context, sourceUri: String) {
        appContext = context.applicationContext
        viewModelScope.launch {
            _state.value = State.Loading
            previewBaseCache = null
            adjustedAlphaCache = null
            strokeLayer.clear()
            activeStrokePoints = null
            try {
                val bitmap = decodePreview(context, Uri.parse(sourceUri))
                    ?: run {
                        _state.value = State.Error(context.getString(R.string.editor_load_failed))
                        return@launch
                    }
                val result = mattingEngine.removeBackground(bitmap, MaskSource.FUSION)
                if (result == null) {
                    _state.value = State.Error(context.getString(R.string.id_photo_matting_failed))
                    return@launch
                }
                // 从 alpha 蒙版提取主体位置，用于「头顶留白」智能构图，避免居中裁剪砍头。
                // 融合 alpha 未锐化，先在参数层默认结果（含默认 sharpen）上提取，与旧行为一致。
                val subject = withContext(Dispatchers.Default) {
                    val sharpened = MaskPostProcessor.adjustEdges(
                        result.alpha, result.width, result.height, EdgeParams()
                    )
                    IDPhotoComposer.subjectBounds(sharpened, result.width, result.height)
                }
                _state.value = State.Ready(
                    originalBitmap = bitmap,
                    alpha = result.alpha,
                    alphaWidth = result.width,
                    alphaHeight = result.height,
                    subject = subject
                )
            } catch (e: Exception) {
                Logger.e(TAG, "IDPhoto load failed", e)
                _state.value = State.Error(
                    context.getString(R.string.editor_load_failed_with_reason, e.message ?: "")
                )
            }
        }
    }

    fun selectColor(index: Int) {
        val current = _state.value as? State.Ready ?: return
        _state.value = current.copy(selectedColorIndex = index)
    }

    fun selectSize(index: Int) {
        val current = _state.value as? State.Ready ?: return
        _state.value = current.copy(selectedSizeIndex = index)
    }

    fun selectTab(tab: IdPhotoTab) {
        val current = _state.value as? State.Ready ?: return
        _state.value = current.copy(activeTab = tab)
    }

    /** 滑块松手时调用（UI 在 onValueChangeFinished 触发，天然防抖）。 */
    fun setEdgeParams(params: EdgeParams) {
        val current = _state.value as? State.Ready ?: return
        _state.value = current.copy(edgeParams = params)
    }

    fun resetEdgeParams() = setEdgeParams(EdgeParams())

    /** 开始一笔涂抹（[radiusPx]/[softness] 已换算为源图像素坐标系）。 */
    fun beginStroke(mode: StrokeMode, radiusPx: Float, softness: Float) {
        if (_state.value !is State.Ready) return
        activeStrokeMode = mode
        activeStrokeRadius = radiusPx
        activeStrokeSoftness = softness
        activeStrokePoints = mutableListOf()
    }

    /** 追加一个源图像素坐标点（UI 经 [IDPhotoComposer.frameToSource] 换算后传入）。 */
    fun appendStrokePoint(point: StrokePoint) {
        activeStrokePoints?.add(point)
    }

    /** 结束一笔：提交描边层，strokeVersion+1 使缓存失效触发底图重建。 */
    fun endStroke() {
        val points = activeStrokePoints ?: return
        activeStrokePoints = null
        if (points.isEmpty()) return
        val current = _state.value as? State.Ready ?: return
        strokeLayer.addStroke(
            BrushStroke(activeStrokeMode, activeStrokeRadius, activeStrokeSoftness, points.toList())
        )
        _state.value = current.copy(
            strokeVersion = current.strokeVersion + 1,
            canUndoStroke = strokeLayer.canUndo,
            canRedoStroke = strokeLayer.canRedo
        )
    }

    /** 是否有进行中的描边（UI 判断是否需要画进行态覆盖层）。 */
    fun hasActiveStroke(): Boolean = activeStrokePoints != null

    fun undoStroke() {
        val current = _state.value as? State.Ready ?: return
        if (!strokeLayer.undo()) return
        _state.value = current.copy(
            strokeVersion = current.strokeVersion + 1,
            canUndoStroke = strokeLayer.canUndo,
            canRedoStroke = strokeLayer.canRedo
        )
    }

    fun redoStroke() {
        val current = _state.value as? State.Ready ?: return
        if (!strokeLayer.redo()) return
        _state.value = current.copy(
            strokeVersion = current.strokeVersion + 1,
            canUndoStroke = strokeLayer.canUndo,
            canRedoStroke = strokeLayer.canRedo
        )
    }

    fun clearStrokes() {
        val current = _state.value as? State.Ready ?: return
        if (strokeLayer.count == 0) return
        strokeLayer.clear()
        _state.value = current.copy(
            strokeVersion = current.strokeVersion + 1,
            canUndoStroke = false,
            canRedoStroke = false
        )
    }

    /**
     * 拖拽/双指缩放微调构图。[dxFraction]/[dyFraction] 为相对预览尺寸的归一化拖拽量
     * （拖动方向与内容一致：向下拖 = 露出更多顶部）；[zoomChange] 为双指缩放比例增量，
     * 缩放范围 [MIN_ZOOM, MAX_ZOOM]，1.0 = cover 填满。
     * 累加后的偏移经 [IDPhotoComposer.clampFraming] 在状态层收敛，拖过边界不累积死区。
     */
    fun transformBy(dxFraction: Float, dyFraction: Float, zoomChange: Float) {
        val current = _state.value as? State.Ready ?: return
        val size = IDPhotoSpecs.SIZES[current.selectedSizeIndex]
        val clamped = IDPhotoComposer.clampFraming(
            srcW = current.originalBitmap.width,
            srcH = current.originalBitmap.height,
            dstW = size.widthPx,
            dstH = size.heightPx,
            framing = framingOf(current).copy(
                offsetX = current.offsetX - dxFraction,
                offsetY = current.offsetY - dyFraction,
                zoom = (current.zoom * zoomChange).coerceIn(MIN_ZOOM, MAX_ZOOM)
            )
        )
        _state.value = current.copy(offsetX = clamped.offsetX, offsetY = clamped.offsetY, zoom = clamped.zoom)
    }

    /** 预览底图缓存键：底色 + 参数层 + 描边版本。手势只改变换参数，不重建底图，保证跟手。 */
    private data class PreviewKey(val colorIndex: Int, val params: EdgeParams, val strokeVersion: Int)

    /** alpha 缓存键：参数层 + 描边版本（alpha 与底色无关，不含 colorIndex）。 */
    private data class AlphaKey(val params: EdgeParams, val strokeVersion: Int)

    @Volatile
    private var previewBaseCache: Pair<PreviewKey, Bitmap>? = null

    @Volatile
    private var adjustedAlphaCache: Pair<AlphaKey, FloatArray>? = null
    private val strokeLayer = StrokeLayer()

    /** 进行中的描边（源图像素坐标，UI 拖完一笔后提交）。 */
    private var activeStrokePoints: MutableList<StrokePoint>? = null
    private var activeStrokeMode: StrokeMode = StrokeMode.RESTORE
    private var activeStrokeRadius: Float = 0f
    private var activeStrokeSoftness: Float = 0f

    /** 参数层 + 描边层叠加后的 alpha（按 AlphaKey 缓存；[strokes] 为 Main 上取的快照）。 */
    private fun adjustedAlpha(current: State.Ready, key: AlphaKey, strokes: List<BrushStroke>): FloatArray {
        adjustedAlphaCache?.takeIf { it.first == key }?.let { return it.second }
        val paramApplied = MaskPostProcessor.adjustEdges(
            current.alpha, current.alphaWidth, current.alphaHeight, current.edgeParams
        )
        val adjusted = StrokeLayer.replay(strokes, paramApplied, current.alphaWidth, current.alphaHeight)
        adjustedAlphaCache = key to adjusted
        return adjusted
    }

    /** 预览底图（original+adjustedAlpha 按当前底色合成，原图尺寸）；按 PreviewKey 缓存，跨手势复用。 */
    suspend fun previewBase(): Bitmap? {
        val current = _state.value as? State.Ready ?: return null
        val key = PreviewKey(current.selectedColorIndex, current.edgeParams, current.strokeVersion)
        val alphaKey = AlphaKey(current.edgeParams, current.strokeVersion)
        val strokes = strokeLayer.snapshot() // Main 上快照，避免 Default 重放与 Main 修改竞态
        return withContext(Dispatchers.Default) {
            previewBaseCache?.takeIf { it.first == key }?.let { return@withContext it.second }
            val base = BackgroundComposer.apply(
                current.originalBitmap, adjustedAlpha(current, alphaKey, strokes),
                current.originalBitmap.width, current.originalBitmap.height,
                IDPhotoSpecs.COLORS[current.selectedColorIndex].argb
            )
            previewBaseCache = key to base
            base
        }
    }

    private fun framingOf(current: State.Ready) = IDPhotoComposer.CropFraming(
        subject = current.subject,
        offsetX = current.offsetX,
        offsetY = current.offsetY,
        zoom = current.zoom
    )

    /** 当前 framing 下的裁剪窗口（纯计算，UI 变换与保存共用同一定位）。 */
    fun currentCropRect(): IDPhotoComposer.CropRect? {
        val current = _state.value as? State.Ready ?: return null
        val size = IDPhotoSpecs.SIZES[current.selectedSizeIndex]
        return IDPhotoComposer.subjectAwareCropRect(
            srcW = current.originalBitmap.width,
            srcH = current.originalBitmap.height,
            dstW = size.widthPx,
            dstH = size.heightPx,
            framing = framingOf(current)
        )
    }

    /** 合成最终输出图（保存用；底图走缓存，仅裁剪+缩放）。 */
    suspend fun composePreview(): Bitmap? = withContext(Dispatchers.Default) {
        val current = _state.value as? State.Ready ?: return@withContext null
        val base = previewBase() ?: return@withContext null
        val size = IDPhotoSpecs.SIZES[current.selectedSizeIndex]
        IDPhotoComposer.cropAndScale(base, size.widthPx, size.heightPx, framingOf(current))
    }

    fun save(context: Context) {
        val current = _state.value as? State.Ready ?: return
        if (current.isSaving) return
        viewModelScope.launch {
            _state.value = current.copy(isSaving = true)
            try {
                val preview = composePreview()
                val outputUri = preview?.let { saveBitmapToMediaStore(context, it) }
                if (outputUri != null) {
                    mediaRepository.refreshMediaLibrary()
                    finishSaving()
                    onSaveComplete?.invoke(outputUri)
                } else {
                    finishSaving(error = context.getString(R.string.editor_save_failed))
                }
            } catch (e: Exception) {
                Logger.e(TAG, "IDPhoto save failed", e)
                finishSaving(
                    error = context.getString(R.string.editor_save_failed_with_reason, e.message ?: "")
                )
            }
        }
    }

    /** 保存结束恢复状态：基于 flow 当前值更新，不吞掉保存期间的拖拽/换色等操作。 */
    private fun finishSaving(error: String? = null) {
        val latest = _state.value as? State.Ready ?: return
        _state.value = latest.copy(isSaving = false, error = error)
    }

    private fun saveBitmapToMediaStore(context: Context, bitmap: Bitmap): String? {
        val size = (_state.value as? State.Ready)?.selectedSizeIndex?.let { IDPhotoSpecs.SIZES[it] }
        val name = "IDPHOTO_${System.currentTimeMillis()}_${size?.widthPx ?: 0}x${size?.heightPx ?: 0}.jpg"
        val values = android.content.ContentValues().apply {
            put(MediaStore.Images.Media.DISPLAY_NAME, name)
            put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg")
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
                put(MediaStore.Images.Media.RELATIVE_PATH, "Pictures/PoLang")
            }
        }
        val uri = context.contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values)
        return uri?.also {
            context.contentResolver.openOutputStream(it)?.use { out ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, JPEG_QUALITY, out)
            }
        }?.toString()
    }

    private fun decodePreview(context: Context, uri: Uri): Bitmap? {
        // 限制解码长边到 DECODE_MAX_DIM：证件照输出仅数百像素，无需原图分辨率；
        // 且可避免原图全尺寸 alpha 合成（BackgroundComposer）在大图上 OOM。
        val resolver = context.contentResolver
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        resolver.openInputStream(uri)?.use { stream -> BitmapFactory.decodeStream(stream, null, bounds) }
        val maxDim = maxOf(bounds.outWidth, bounds.outHeight).coerceAtLeast(1)
        val sample = if (maxDim > DECODE_MAX_DIM) maxDim / DECODE_MAX_DIM else 1
        val opts = BitmapFactory.Options().apply { inSampleSize = sample }
        return resolver.openInputStream(uri)?.use { stream -> BitmapFactory.decodeStream(stream, null, opts) }
    }

    var onSaveComplete: ((String) -> Unit)? = null

    override fun onCleared() {
        super.onCleared()
        (mattingEngine as? MattingEngineImpl)?.release()
    }
}

File: iosApp/PoLang/Features/IdPhoto/IDPhotoViewModel.swift
import CoreGraphics
import Foundation
import Photos
import UIKit

// MARK: - 证照屏 ViewModel（对标 specs/screens/idphoto.yaml §7 状态机与数据流）
// UI 层契约见 IDPhotoScreen；纯逻辑在 IdPhotoDomain；推理在 MattingEngine。
// 线程契约：状态拷贝在 Main；全图计算经串行 compute 队列（描边快照在进队前取）；
// 模型下载中心交互（路径解析/enqueue）在 Main。

enum IdPhotoState {
    case loading
    case error(String)

    struct Ready {
        var selectedColorIndex: Int = 0
        var selectedSizeIndex: Int = 0
        var activeTab: IdPhotoTab = .bgColor
        var edgeParams: EdgeParams = .defaultValue
        var strokeVersion: Int = 0
        var canUndoStroke: Bool = false
        var canRedoStroke: Bool = false
        var hasStrokes: Bool = false
        var isSaving: Bool = false
    }

    case ready(Ready)
}

@MainActor
final class IDPhotoViewModel: ObservableObject {

    // 常量契约
    private static let decodeMaxDim = 1024
    private static let jpegQuality: CGFloat = 0.95
    private static let minZoom: Float = 1.0
    private static let maxZoom: Float = 4.0

    @Published private(set) var state: IdPhotoState = .loading

    var onSaved: (() -> Void)?
    var onSaveFailed: ((String) -> Void)?

    private let localIdentifier: String
    private let engine = IDPhotoMattingEngine()

    // Ready 数据（非 UI 投影部分）
    private var original: CGImage?
    private var originalPixels: [UInt8] = []
    private var sourceW = 0
    private var sourceH = 0
    private var rawAlpha: [Float] = []
    private var subject: (top: Int, centerX: Float)?
    private var offsetX: Float = 0
    private var offsetY: Float = 0
    private var zoom: Float = 1

    // 描边栈
    private var strokes: [BrushStroke] = []
    private var redoStack: [BrushStroke] = []
    private var activeStroke: BrushStroke?

    // 缓存（key = 内容指纹）
    private struct AdjustKey: Equatable {
        var edgeParams: EdgeParams
        var strokeVersion: Int
    }
    private struct PreviewKey: Equatable {
        var colorIndex: Int
        var edgeParams: EdgeParams
        var strokeVersion: Int
    }
    private var adjustedAlphaCacheKey: AdjustKey?
    private var adjustedAlphaCache: [Float] = []
    private var previewCacheKey: PreviewKey?
    private var previewCache: CGImage?

    /// 串行计算队列：掩码变换/合成/推理全走这里（重放前快照在 Main 已取）
    private let computeQueue = DispatchQueue(label: "com.mamba.picme.idphoto.compute", qos: .userInitiated)

    private func runCompute<T>(_ work: @escaping () throws -> T) async throws -> T {
        try await withCheckedThrowingContinuation { continuation in
            computeQueue.async {
                do {
                    continuation.resume(returning: try work())
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    init(localIdentifier: String) {
        self.localIdentifier = localIdentifier
        Task { await load() }
    }

    deinit {
        engine.release()
    }

    // MARK: - 加载

    private func load() async {
        guard let uiImage = await ThumbnailLoader.shared.fullResolution(for: localIdentifier),
              let fullCg = uiImage.cgImage else {
            state = .error(L("editor_load_failed"))
            return
        }
        // 长边 ≤1024 降采样（契约 DECODE_MAX_DIM）
        let cg = IdPhotoBitmap.downscale(fullCg, maxDim: Self.decodeMaxDim)
        // Main 侧预解析 modnet 路径（下载中心属 MainActor；缺失 → enqueue 下载，本次报错重进重试）
        let modnetPath = Self.resolveModnetPath()
        do {
            let result = try await runCompute { [engine] in
                try engine.removeBackground(cg, modnetModelPath: modnetPath)
            }
            guard let buffer = IdPhotoBitmap.rgbaBuffer(from: cg) else {
                state = .error(L("editor_load_failed"))
                return
            }
            let bounds = try? await runCompute {
                IDPhotoComposer.subjectBounds(result.alpha, w: result.width, h: result.height)
            }
            original = cg
            originalPixels = buffer.pixels
            sourceW = buffer.width
            sourceH = buffer.height
            rawAlpha = result.alpha
            subject = bounds
            offsetX = 0
            offsetY = 0
            zoom = 1
            adjustedAlphaCacheKey = nil
            previewCacheKey = nil
            state = .ready(IdPhotoState.Ready())
        } catch MattingError.modelMissing(let modelId) {
            // 引擎已 enqueue 下载；本次报错，返回重进即重试（Android 对齐）
            state = .error(String(format: L("editor_load_failed_with_reason"), modelId))
        } catch {
            state = .error(L("id_photo_matting_failed"))
        }
    }

    // MARK: - 面板操作（Main 同步状态拷贝）

    func selectColor(_ index: Int) {
        updateReady { $0.selectedColorIndex = index }
    }

    func selectSize(_ index: Int) {
        updateReady { $0.selectedSizeIndex = index }
    }

    func selectTab(_ tab: IdPhotoTab) {
        updateReady { $0.activeTab = tab }
    }

    func setEdgeParams(_ params: EdgeParams) {
        updateReady { $0.edgeParams = params }
    }

    func resetEdgeParams() {
        updateReady { $0.edgeParams = .defaultValue }
    }

    // MARK: - 构图手势

    /// dxFraction/dyFraction：拖动位移/预览框尺寸（正=手指右/下移）；zoomChange：缩放增量因子。
    /// 语义：内容跟手——offsetX -= dx（右拖 → 取景窗左移）；随后 clampFraming 收敛防死区。
    func transformBy(dxFraction: Float, dyFraction: Float, zoomChange: Float) {
        guard case .ready = state else { return }
        let spec = currentSizeSpec()
        let clamped = IDPhotoComposer.clampFraming(
            subject: subject,
            offsetX: offsetX - dxFraction, offsetY: offsetY - dyFraction,
            zoom: zoom * zoomChange,
            srcW: sourceW, srcH: sourceH,
            dstW: spec.pixelW, dstH: spec.pixelH
        )
        offsetX = clamped.offsetX
        offsetY = clamped.offsetY
        zoom = min(Self.maxZoom, max(Self.minZoom, clamped.zoom))
    }

    /// 当前取景窗（源图像素）；供预览绘制与坐标换算
    func currentCropRect() -> CropRect? {
        guard case .ready(let ready) = state else { return nil }
        let spec = IDPhotoSizeSpec.allCases[ready.selectedSizeIndex]
        return IDPhotoComposer.subjectAwareCropRect(
            subject: subject,
            offsetX: offsetX, offsetY: offsetY, zoom: zoom,
            srcW: sourceW, srcH: sourceH,
            dstW: spec.pixelW, dstH: spec.pixelH
        )
    }

    // MARK: - 描边（快照契约：Main 取快照 → compute 重放）

    func beginStroke(mode: StrokeMode, radiusSourcePx: Float, softness: Float) {
        activeStroke = BrushStroke(mode: mode, radiusPx: max(1, radiusSourcePx),
                                   softness: softness, points: [])
    }

    func appendStrokePoint(xSourcePx: Float, ySourcePx: Float) {
        activeStroke?.points.append(StrokePoint(x: xSourcePx, y: ySourcePx))
    }

    func endStroke() {
        guard var stroke = activeStroke else { return }
        stroke.points = stroke.points.count == 1
            ? [stroke.points[0], stroke.points[0]]
            : stroke.points
        activeStroke = nil
        guard !stroke.points.isEmpty else { return }
        strokes.append(stroke)
        redoStack.removeAll()
        bumpStrokeState()
    }

    var hasActiveStroke: Bool { activeStroke != nil }

    func undoStroke() {
        guard let last = strokes.popLast() else { return }
        redoStack.append(last)
        bumpStrokeState()
    }

    func redoStroke() {
        guard let stroke = redoStack.popLast() else { return }
        strokes.append(stroke)
        bumpStrokeState()
    }

    func clearStrokes() {
        guard !strokes.isEmpty || !redoStack.isEmpty else { return }
        strokes.removeAll()
        redoStack.removeAll()
        bumpStrokeState()
    }

    private func bumpStrokeState() {
        updateReady {
            $0.strokeVersion += 1
            $0.canUndoStroke = !strokes.isEmpty
            $0.canRedoStroke = !redoStack.isEmpty
            $0.hasStrokes = !strokes.isEmpty
        }
    }

    // MARK: - 底图合成（缓存 key = color + edge + strokeVersion）

    /// 合成底图；nil = 构建中/失败（UI 保留上一帧防闪白）
    func previewBase() async -> CGImage? {
        guard case .ready(let ready) = state, original != nil else { return nil }
        let key = PreviewKey(colorIndex: ready.selectedColorIndex,
                             edgeParams: ready.edgeParams,
                             strokeVersion: ready.strokeVersion)
        if let cached = previewCache, previewCacheKey == key {
            return cached
        }
        // Main 取快照（线程契约）
        let colorIndex = ready.selectedColorIndex
        let edgeParams = ready.edgeParams
        let strokeSnapshot = strokes
        let raw = rawAlpha
        let pixels = originalPixels
        let w = sourceW
        let h = sourceH
        let cg = (try? await runCompute { () -> CGImage? in
            var alpha = raw
            alpha = MaskPostProcessor.adjustEdges(alpha, w: w, h: h, params: edgeParams)
            alpha = StrokeLayer.replay(strokes: strokeSnapshot, base: alpha, w: w, h: h)
            guard IDPhotoColorSpec.allCases.indices.contains(colorIndex) else { return nil }
            let color = IDPhotoColorSpec.allCases[colorIndex].rgb
            let composed = BackgroundComposer.composeOnColor(pixels: pixels, alpha: alpha, bgColor: color)
            return IdPhotoBitmap.cgImage(from: composed, width: w, height: h)
        })
        guard let image = cg else { return nil }
        previewCacheKey = key
        previewCache = image
        return image
    }

    // MARK: - 保存（WYSIWYG：与预览同 base 同 cropRect）

    func composePreview() async -> CGImage? {
        guard let base = await previewBase(),
              let crop = currentCropRect(),
              let cropped = base.cropping(to: CGRect(x: crop.left, y: crop.top,
                                                     width: crop.width, height: crop.height)) else {
            return nil
        }
        let spec = currentSizeSpec()
        // 精确拉伸到规格像素尺寸（与预览框渲染同语义：cover 零留边）
        var pixels = [UInt8](repeating: 0, count: spec.pixelW * spec.pixelH * 4)
        return pixels.withUnsafeMutableBytes { ptr -> CGImage? in
            guard let ctx = CGContext(
                data: ptr.baseAddress,
                width: spec.pixelW, height: spec.pixelH,
                bitsPerComponent: 8, bytesPerRow: spec.pixelW * 4,
                space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
            ) else { return nil }
            ctx.interpolationQuality = .high
            ctx.draw(cropped, in: CGRect(x: 0, y: 0, width: spec.pixelW, height: spec.pixelH))
            return ctx.makeImage()
        }
    }

    func save() async {
        guard case .ready(var ready) = state, !ready.isSaving else { return }
        ready.isSaving = true
        state = .ready(ready)
        defer {
            // 读最新 state 复位（不吞保存期间的操作）
            if case .ready(var latest) = state {
                latest.isSaving = false
                state = .ready(latest)
            }
        }
        guard let cg = await composePreview() else {
            onSaveFailed?(L("editor_save_failed"))
            return
        }
        // JPEG 0.95 契约（PHAsset 由系统命名——双端差异已登记 spec §12）
        let image = UIImage(cgImage: cg)
        guard let jpeg = image.jpegData(compressionQuality: Self.jpegQuality),
              let finalImage = UIImage(data: jpeg) else {
            onSaveFailed?(L("editor_save_failed"))
            return
        }
        do {
            try await PhotoSaver.saveToLibrary(finalImage)
            onSaved?()
        } catch {
            onSaveFailed?(L("editor_save_failed"))
        }
    }

    // MARK: - 工具

    /// Main 侧解析 modnet 模型路径；缺失时 enqueue 下载（Android initializeWithDownloadFallback 对齐）
    private static func resolveModnetPath() -> String? {
        let manager = ModelDownloadManager.shared
        let path = manager.modelsDir
            .appendingPathComponent(IDPhotoMattingEngine.modnetModelId)
            .appendingPathComponent("modnet.onnx").path
        guard FileManager.default.fileExists(atPath: path) else {
            if !manager.isModelDownloaded(IDPhotoMattingEngine.modnetModelId) {
                manager.download(IDPhotoMattingEngine.modnetModelId)
            }
            return nil
        }
        return path
    }

    private func currentSizeSpec() -> IDPhotoSizeSpec {
        guard case .ready(let ready) = state,
              IDPhotoSizeSpec.allCases.indices.contains(ready.selectedSizeIndex) else {
            return .in1
        }
        return IDPhotoSizeSpec.allCases[ready.selectedSizeIndex]
    }

    private func updateReady(_ mutate: (inout IdPhotoState.Ready) -> Void) {
        guard case .ready(var ready) = state else { return }
        mutate(&ready)
        state = .ready(ready)
    }
}
